radicale/storage.py

```diff
@@ -621,43 +689,150 @@ def move(cls, item, to_collection, to_href):
-            self.logger.debug(
-                "Can't translate name safely to filesystem: %s", href)
-            return None
-        path = path_to_filesystem(self._filesystem_path, href)
-        if not os.path.isfile(path):
-            return None
-        with open(path, encoding=self.encoding, newline="") as f:
-            text = f.read()
+    _item_cache_cleaned = False
+
+    def get(self, href, verify_href=True):
+        item, metadata = self._get_with_metadata(href, verify_href=verify_href)
+        return item
+
+    def _get_with_metadata(self, href, verify_href=True):
+        # Like ``get`` but additonally returns the following metadata:
+        # tag, start, end: see ``xmlutils.find_tag_and_time_range``
+        if verify_href:
+            try:
+                if not is_safe_filesystem_path_component(href):
+                    raise UnsafePathError(href)
+                path = path_to_filesystem(self._filesystem_path, href)
+            except ValueError:
+                self.logger.debug(
+                    "Can't translate name safely to filesystem: %s", href)
+                return None, None
+        else:
+            path = os.path.join(self._filesystem_path, href)
+        try:
+            with open(path, "rb") as f:
+                btext = f.read()
+        except (FileNotFoundError, IsADirectoryError):
+            return None, None
+        # The hash of the component in the file system. This is used to check,
+        # if the entry in the cache is still valid.
+        input_hash = md5()
+        input_hash.update(btext)
+        input_hash = input_hash.hexdigest()
+        cache_folder = os.path.join(self._filesystem_path, ".Radicale.cache",
+                                    "item")
+        try:
+            with open(os.path.join(cache_folder, href), "rb") as f:
+                cinput_hash, cetag, ctext, ctag, cstart, cend = pickle.load(f)
+        except FileNotFoundError:
+            cinput_hash = cetag = ctext = ctag = cstart = cend = None
+        vobject_item = None
+        if input_hash != cinput_hash:
+            vobject_item = Item(self, href,
+                                text=btext.decode(self.encoding)).item
+            ctext = vobject_item.serialize()
+            cetag = get_etag(ctext)
+            ctag, cstart, cend = xmlutils.find_tag_and_time_range(vobject_item)
+            self._makedirs_synced(cache_folder)
+            try:
+                # Race: Other processes might have created and locked the
+                # file.
+                with self._atomic_write(os.path.join(cache_folder, href),
+                                        "wb") as f:
+                    pickle.dump((input_hash, cetag, ctext,
+                                 ctag, cstart, cend), f)
+            except PermissionError:
+                pass
+            # Clean cache entries (max once per request)
+            if not self._item_cache_cleaned:
```

Comment:
This only happens once, after new items are uploaded to the storage. I put it here and not into ``def upload(...)`` because the user can also edit the file system storage by hand.